ipfix-flow-exporter/bpf_ipfix_egress_kern.c

```diff
@@ -249,13 +249,21 @@ void parse_ipv4(struct __sk_buff *skb, u64 l3_offset)
     u16 control_bit = 0;
     u16 icmp_type = 0;
 
-    if (iph + 1 > data_end)
+    if (iph->ihl < 5 || iph + 1 > data_end)
         return;
 
+    u8 l4_offset = iph->ihl * 4; // ip header length
+ 
+    iphdr *thdr = iph + 1; // transport layer header pointer
+
+    //  Check if there are any options present
+    if (l4_offset > sizeof(struct iphdr *)) 
+        thdr = (struct iphdr *)((unsigned char *)iph + l4_offset);
```

Comment:
No need to have if condition for this. Just initialize thdr to ((unsigned char *)iph + l4_offset), which will work for all the cases.

Reply:
thdr is not of type (struct iphdr *) It is the start of Transport header which is either TCP or UDP. You can define a void pointer as parse_icmp_type() & parse_port() functions accept void pointer as arg.